### PR TITLE
[WIP] Implement UIA Invoke Method

### PR DESCRIPTION
**Pull request by chiaramooney**
### Type of Change
- New feature (non-breaking change which adds functionality)

### Why
Implement UIA Invoke method. Method is used by automation clients to invoke the action of pressable, stateless controls. 

### What
Implement UIA Invoke method. 

Note current implementation is not control specific. i.e. the implementation applies to all controls that CompositionDynamicAutomationPeer controls. To match Paper behavior, I believe we will need to only implement the invoke method for controls that are pressable and stateless. Also, the implementation invokes both the onClick and onAccessibilityTap functionality when the method is called. To match Paper behavior, this will need to be adjusted so that the onAccessibilityTap method is only called when onClick is not implemented. 

## Testing
Programmatically called method to confirm method performs intended function of carrying out the action of the control upon press. 

TODO
- [x] Test that method works when called by an automation client. 
 ###### Microsoft Reviewers: codeflow:open?pullrequest=https://github.com/microsoft/react-native-windows/pull/11874&drop=dogfoodAlpha

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (FalseLobster) [inline vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:286]: I do wonder if this is correct? What if we have some special code that needs to behave differently for pointer and AT users? (e.g. Telemetry recording invocation method).

I feel like if both are implemented, maybe we should only call onAccessibilityTap and rely on users to handle the normal onClick behavior in the JS callback. Or maybe both? Maybe this was already discussed with the Paper impl, but I don't know that I like the idea that we deprive people of an AT-specific code path.

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (FalseLobster) [inline vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:290]: May be worth noting we need to hook up accessibility action here as well: https://reactnative.dev/docs/accessibility#accessibility-actions

Could probably be done in a separate PR though

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (FalseLobster) [inline vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:292]: This seems incorrect. We should be raising the UIA Invoke event on this View not the root CV.

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (FalseLobster) [inline vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:275]: You also need to return this object in GetPatternProvider above (line 126) I believe

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionRootAutomationProvider.h
Comment (FalseLobster) [inline vnext/Microsoft.ReactNative/Fabric/Composition/CompositionRootAutomationProvider.h:18]: Is this necessary? The implementation does nothing and I'm not sure people are going to be invoking their root views.

**Review by FalseLobster — CHANGES_REQUESTED** [review]

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionRootAutomationProvider.h
Comment (chiaramooney) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionRootAutomationProvider.h:18]: That makes sense! Will change!

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (chiaramooney) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:303]: For sure! Just to confirm I understand what you're saying, you're wondering that maybe the behavior we had on Paper was wrong. And that instead we should either...
(1) always call onAccessibilityTap
(2) always call both onAccessibilityTap and onClick

That makes sense to me! It did seem odd that the onAccessibilityTap functionality would be ignored when onClick functionality is set. Option 2 seems like the right option to me, do you have an instinct?

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (FalseLobster) [inline vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:127]: A few issues with this function:

1. Regardless of what you're assigning to \*pRetVal on 142, it immediately gets null-ed out on 145.  A common pattern in COM out params (the IUnknown**) is to check if the pointer-pointer is valid (lines 127 & 128) and then initialize the pointer to null so you can early return if you don't have a return value.  Basically, move line 145 to 129.
2. I don't think it's good practice to upcast the this pointer on 142 to IUnknown using reinterpret_cast. I'm rusty on C++ from all the TS I do these days; but since we're coming from a derived type, I thought this was implicit? But reinterpret_cast certainly seems off.
3. I think you might be scoping the Invoke pattern a little too aggressively on 141, it's a pretty common pattern. I would expect to see [menuitem](https://learn.microsoft.com/en-us/windows/win32/winauto/uiauto-supportmenuitemcontroltype#required-control-patterns) and [link](https://learn.microsoft.com/en-us/windows/win32/winauto/uiauto-supporthyperlinkcontroltype#required-control-patterns) here, too, for example.  For controls like MenuItem where it says "Depends" we might need to also check to see if they're supplying a onAccessibilityTap/AccessibilityAction callback to know opting into the pattern is intended.

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (FalseLobster) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:301]: I just checked and this:
> maybe we should only call onAccessibilityTap and rely on users to handle the normal onClick behavior in the JS callback

is what we do in Office, though I do want to get this right.  So while (2) sounds reasonable, I wouldn't mind a second opinion here from @jonthysell or @acoates-ms

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (chiaramooney) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp]: Resolved now!

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:311]: I'd go with (1).  I dont see any indication in code that core calls onClick when invoking onAccessibilityTap.  @Saadnajmi might be able to verify the behavior on iOS.  (I think onClick might be fabric only in iOS)

Comment (jonthysell) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:311]: It appears that `onAccessibilityTap` is its own, independent event in RN code, and technically an iOS-only event (though not documented that way, see https://github.com/facebook/react-native/issues/34136).

See also https://reactnative.dev/docs/view#onaccessibilitytap

I think our Paper impl is too tied up with how we interacted with XAML and XAML's behaviors. `onClick` is a windows-only event, and I don't really like that we assume to send and `onClick` when hit via UIA. Why not an `onPress`? Nowhere do I see code that defaults to calling RN's own built-in `onPress` during an `onAccessibilityTap`.

I'd rather leave that up to app devs in JS rather than assume a default behavior "by default". Every example online I can find people who care about accessibility define both onPress and onAccessibilityTap (even if it's usually to the same function).

That said, it appears in RN for iOS, there is a kind of event chaining where the 'activate' accessibilityAction presumes to call `onAccessibilityTap` if it's available: 

https://github.com/facebook/react-native/blob/8d8455c3c34e9554908781485718d8eee252267f/packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm#L802-L810

```mm
- (BOOL)accessibilityActivate
{
  if (_eventEmitter && _props->onAccessibilityTap) {
    _eventEmitter->onAccessibilityTap();
    return YES;
  } else {
    return NO;
  }
}
```

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:142]: I think it's possible we could/should decouple this from `accessibilityRole`. Regardless of role, if we chould just check if `onAccessibilityTap` exists on the element, and if it does, return we're IInvokable, otherwise, not invokable. Thoughts?

File: vnext/ReactCommon/TEMP_UntilReactCommonUpdate/react/renderer/components/view/ViewEventEmitter.cpp
Comment (jonthysell) [inline vnext/ReactCommon/TEMP_UntilReactCommonUpdate/react/renderer/components/view/ViewEventEmitter.cpp:160]: Do we even need onClick when Pressability is a thing now? https://reactnative.dev/docs/pressable

File: change/@react-native-windows-cli-9a018e15-0273-415e-959f-3e4be94e544c.json
Comment (jonthysell) [inline change/@react-native-windows-cli-9a018e15-0273-415e-959f-3e4be94e544c.json:1]: These don't look like they belong here. When was the last time you merged main?

File: change/react-native-windows-6ad43580-2e16-403a-b4f2-1238f29863b5.json
Comment (jonthysell) [inline change/react-native-windows-6ad43580-2e16-403a-b4f2-1238f29863b5.json:1]: These don't look like they belong here. When was the last time you merged main?

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (chiaramooney) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:311]: Couple of additional pieces of info:

- Upstream has recently added a topClick/onClick event, so it won't be Windows only anymore. See https://github.com/microsoft/react-native-windows/pull/11969. More of a fyi than something actionable on this PR. 
- The value of `props->onClick` matches whether `onPress` has been set. And the call to  `dispatchEvent("topClick")` runs the `onPress` function. I haven't dug into the code to figure out why this is the case, but that's the behavior we get. 
- From testing, it does NOT appear that if `onAccessibilityTap` is not set, then the `onPress` function is run instead. The program simply does a noop. So if the Invoke method expects the behavior to match when you click without AT, then just calling `onAccessibilityTap` might not be sufficient.
- Based on the RN docs, I believe we'll want to implement the accessibilityAction "activate" to match whatever behavior we land on for Invoke. They are both described in the same way. What's interesting is that with this description "Typically this should perform the same action as when the user touches or clicks the component when not using an assistive technology. This is generated when a screen reader user double taps the component." I would expect `_props->onClick` to be called if `onAccessibilityTap` isn't implemented.

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (chiaramooney) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:142]: Couple of thoughts:
- In the UIA documentation the rules around UIA providers are very specific to each control. i.e. just because a control can be pressed doesn't mean Invokable is the right provider. The control might instead need to implement the Toggle or Selection provider. I think if we were to make this change it would put a lot of ownership on the developer to setup their controls correctly. For example, all button controls should support the Invoke provider according to UIA. But with this change there could be buttons that don't implement onAccessibilityTap that then would not support the Invoke pattern.
- The `onAccessibilityTap` prop can be specified by any control. So, with this change, developers would also now be able to make controls Invokable that shouldn't have that UIA pattern. i.e. You could specify onAccessibilityTap on a Switch control, and it would have the Invoke pattern when it should only support the Toggle pattern. 

UIA Reference: https://learn.microsoft.com/en-us/windows/win32/winauto/uiauto-supportbuttoncontroltype

File: change/@react-native-windows-cli-9a018e15-0273-415e-959f-3e4be94e544c.json
Comment (chiaramooney) [reply change/@react-native-windows-cli-9a018e15-0273-415e-959f-3e4be94e544c.json:1]: Yea those might be stale. I'll refresh before merging.

File: change/react-native-windows-6ad43580-2e16-403a-b4f2-1238f29863b5.json
Comment (chiaramooney) [reply change/react-native-windows-6ad43580-2e16-403a-b4f2-1238f29863b5.json:1]: Yea those might be stale. I'll refresh before merging.

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:311]: > "Typically this should perform the same action as when the user touches or clicks the component when not using an assistive technology. This is generated when a screen reader user double taps the component." 

> I would expect `_props->onClick` to be called if `onAccessibilityTap` isn't implemented.

I think that it is a note to the developer that they should typically set `onAccessibilityTap` to the same value of `onClick`.  Not that the platform would do it for you.  If the platform does it, then it is no longer **typically**, but **always**, which could be problematic for some use cases.

File: vnext/ReactCommon/TEMP_UntilReactCommonUpdate/react/renderer/components/view/ViewEventEmitter.cpp
Comment (chiaramooney) [reply vnext/ReactCommon/TEMP_UntilReactCommonUpdate/react/renderer/components/view/ViewEventEmitter.cpp:160]: Copying note from above: The value of props->onClick matches whether onPress has been set. And the call to dispatchEvent("topClick") runs the onPress function. I haven't dug into the code to figure out why this is the case, but that's the behavior we get.

So we don't need to pass any additional props through on the JS side to get onClick data. This code is just to allow us to access the onPress data on the native side.

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (chiaramooney) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:311]: I see that makes sense. Seems consensus is to only calling onAccessibilityTap. I'll make those edits.

**Comment by chiaramooney** [comment]
@jonthysell @acoates-ms @FalseLobster

File: vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp
Comment (FalseLobster) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:142]: > I think if we were to make this change it would put a lot of ownership on the developer to setup their controls correctly

I think for JS-based controls written from the ground up, there needs to be ownership on the dev's part. We should document how these things interact and provide guidance, ideally, but we should also provide our own control offerings via FURN that solve these problems for them so they don't have to think about them as much.

My 2c:  

1. If we have something where Invoke MUST be present (e.g. Link), we should just look at the Role and expose the pattern.  It's then a bug on the JS dev if they fail to provide onAccessibilityTap/onAccessibilityAction(activate).
2. If it's something like Button or MenuItem where it can be Invoke/Toggle/ExpandCollapse in mutual exclusivity, we need to look at role & the provided accessibiliyAction, since the docs support extending the actions for custom types we'll probably want "toggle", "expand", "collapse", to make the decision what pattern to expose.  In this case it'd be a bug on JS dev to supply multiple patterns (both "activate" and "toggle")

Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/Fabric/Composition/CompositionDynamicAutomationProvider.cpp:142]: Some of those sound like things we could possibly make the JS start creating yellowboxes for.

**Review by FalseLobster — APPROVED** [review]

**Pull request merged**
